server/master/src/main/java/org/apache/accumulo/master/metrics/fate/FateMetrics.java

```diff
@@ -16,61 +16,234 @@
+    }
+  }
+
+  /**
+   * For testing only: allow refresh delay to be set to any value, over riding the enforced minimum.
+   *
+   * @param minimumRefreshDelay
+   *          set new min refresh value, in seconds.
+   */
+  void overrideRefresh(final long minimumRefreshDelay) {
+    long delay = Math.max(0, minimumRefreshDelay);
+    this.minimumRefreshDelay = TimeUnit.SECONDS.toMillis(delay);
   }
 
   @Override
   protected void prepareMetrics() {
-    FateMetricValues fateMetrics = metricValues.get();
+
     long now = System.currentTimeMillis();
+
     if ((lastUpdate + minimumRefreshDelay) < now) {
-      fateMetrics = FateMetricValues.updateFromZookeeper(context, fateMetrics);
-      metricValues.set(fateMetrics);
+      metricValues.set(updateFromZookeeper());
+
       lastUpdate = now;
-      // update individual gauges that are reported.
-      currentFateOps.set(fateMetrics.getCurrentFateOps());
-      zkChildFateOpsTotal.set(fateMetrics.getZkFateChildOpsTotal());
-      zkConnectionErrorsTotal.set(fateMetrics.getZkConnectionErrors());
+
+      recordValues();
     }
   }
 
+  private void recordValues() {
+
+    FateMetricValues measurement = metricValues.get();
+
+    // update individual gauges that are reported.
+    currentFateOps.set(measurement.getCurrentFateOps());
+    zkChildFateOpsTotal.set(measurement.getZkFateChildOpsTotal());
+    zkConnectionErrorsTotal.set(measurement.getZkConnectionErrors());
+
+    // the number FATE Tx states (NEW< IN_PROGRESS...) are fixed - the underlying
+    // getTxStateCounters call will return a current valid count for each possible state.
+    Map<String,Long> states = measurement.getTxStateCounters();
+
+    states.forEach((key, value) -> {
+      MutableGaugeLong v = fateTypeCounts.get(key);
+      if (v != null) {
+        v.set(value);
+      } else {
+        v = super.getRegistry().newGauge(metricNameHelper(FATE_TX_STATE_METRIC_PREFIX, key),
+            "By transaction state count for " + key, value);
+        fateTypeCounts.put(key, v);
+      }
+    });
+
+    // the op types are dynamic and the metric gauges generated when first seen. After
+    // that the values need to be cleared and set any new values present. This is so
+    // that the metrics system will report "known" values once seen. In operation, the
+    // number of types will be a fairly small set and should populate with normal operations.
+
+    // clear current values.
+    fateOpCounts.forEach((key, value) -> value.set(0));
+
+    // update new counts, create new gauge if first time seen.
+    Map<String,Long> opTypes = measurement.getOpTypeCounters();
+
+    opTypes.forEach((key, value) -> {
+      MutableGaugeLong g = fateOpCounts.get(key);
+      if (g != null) {
+        g.set(value);
+      } else {
+        g = super.getRegistry().newGauge(metricNameHelper(FATE_OP_TYPE_METRIC_PREFIX, key),
+            "By transaction op type count for " + key, value);
+      }
+
+      fateOpCounts.put(key, g);
```

Comment:
I'll move the `fateOpCounts.put(key.g)`  so that only created instances are added to the map.  The set in not necessary on gauge creation, it is set to the initial value in the` newGauge(...)` method.